llmfoundry/models/hf/hf_fsdp.py

```diff
@@ -136,16 +148,39 @@ def prepare_hf_causal_lm_model_for_fsdp(model: PreTrainedModel) -> None:
                 f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
                 'follow common layer/weight naming conventions.')
     block_type = type(model_block[0])  # type: ignore
-    # When using the HF LM models,
-    # the weights of the self.lm_head and self.transformer.wte are tied.
-    # This tying occurs inside the `self.post_init()` function.
-    # This is a hurdle for FSDP because they need to be in the same FSDP block
-    # These lines ensures that both modules stay together in the top-most block when
-    # the model has this tying enabled (almost all do; this property defaults to True)
-    if model.config.tie_word_embeddings:
-        causal_base_model._fsdp_wrap = False  # type: ignore
-        tied_embeddings._fsdp_wrap = False  # type: ignore
-        lm_head._fsdp_wrap = False  # type: ignore
+    if init_device == 'mixed':
+        # For FSDP with models with different device intiailizations, `mixed`, which
+        # initializes the model on rank 0 on `cpu` and on all other ranks on `meta,``
+        # we need to tag all child modules that are torch.nn.Modules with `_fsdp_wrap`.
+        for child in model.children():
+            if isinstance(child, type(causal_base_model)):
+                continue
+            if isinstance(child, torch.nn.Module):
+                child._fsdp_wrap = True
+
+        for child in causal_base_model.children():
+            if isinstance(child, torch.nn.ModuleList):
+                continue
+            if isinstance(child, torch.nn.Module):
+                child._fsdp_wrap = True
+
+        if model.config.tie_word_embeddings:
+            warnings.warn((
+                'The passed in HuggingFaceModel has tied word embeddings '
+                'and the passed in initialization device is `mixed.` '
+                'In order to support this initializaiton schema, we are going to '
+                'break weight tying.'))
```

Comment:
I strongly think this should be an error. Our warnings are way too easy to miss and IMO someone should have to explicitly set the config overrides to have `tie_word_embeddings: False`.
If it is made to be an error, we might also want to expand the check a bit because MPT has `tie_word_embeddings: True` but the tie won't break because `tied_embeddings == lm_head`. I _think_ that as long as those point to the same module they will, by definition, be on the same FSDP block and the weight tying won't be (perhaps even couldn't be?) broken.

Reply:
I agree, changed to an error...